src/psnr.jl

```diff
@@ -0,0 +1,67 @@
+"""
+    PeakSignalNoiseRatio <: FullReferenceIQI
+    psnr(x, ref [, peakval])
+    assess(PSNR(), x, ref, [, peakval])
+
+Peak signal-to-noise ratio (PSNR) is used to measure the quality of image.
+
+PSNR (in dB) is calculated by `10log10(PeakVal^2/mse(x, ref))`, where `PeakVal`
+is the maximum possible pixel value of image `ref`, and
+[`mse`](@ref ImageDistances.mse) is the mean squared error between images `x`
+and `ref`.
+
+!!! info
+
+    For general `Color3` images, PSNR is reported against each channel of that
+    color space as a `Vector`. One exception is `AbstractRGB` images, where
+    `mse` is calculated over three channels by convention.
+
+!!! info
+
+    `peakval` is optional only for gray and RGB color space. For other color
+    spaces, you need to explicitly pass it as a `Vector`.
+"""
+struct PeakSignalNoiseRatio <: FullReferenceIQI end
+const PSNR = PeakSignalNoiseRatio # alias PSNR since it's too famous
+
+# api
+(iqi::PSNR)(x, ref, peakval) = psnr(x, ref, peakval)
+(iqi::PSNR)(x, ref) = iqi(x, ref, peak_value(eltype(ref)))
+
+@doc (@doc PeakSignalNoiseRatio)
+psnr(x, ref, peakval) = psnr(x, ref, peakval)
+psnr(x, ref) = psnr(x, ref, peak_value(eltype(ref)))
+
+
+# implementation
+""" Define the default peakval for colors """
+peak_value(::Type{T}) where T <: NumberLike = one(eltype(T))
+peak_value(::Type{T}) where T <: AbstractRGB = one(eltype(T))
+function peak_value(::Type{T}) where T <: Color3
+    err_msg = "peakval for PSNR can't be inferred and should be explicitly passed for $(T) images"
+    throw(ArgumentError(err_msg))
+end
+
+psnr(x::GenericGrayImage, ref::GenericGrayImage, peakval::Real) =
+    20log10(peakval) - 10log10(mse(x, ref))
+
+psnr(x::GenericImage{<:AbstractRGB}, ref::GenericImage{<:AbstractRGB},
```

Comment:
`psnr` becomes meaningful only in cases that `x` and `ref` are the same type. E.g., `RGB`-`RGB`.

The problem with this annotation is that this function will not throw error for `RGB`-`BGR`.

A solution is to convert `x` to `typeof(ref)`. Will fix later.

Reply:
In the new commit,`x` will be converted to type of `ref` when necessary.
E.g., `psnr(Lab.(x), ref) == psnr(RGB.(Lab.(x)), ref)`